P2: Usage examples moved to appropriate pages (#624)
mongodb/docs-java: source/logging-monitoring/change-streams.txt

   see the :ref:`<replica-set-oplog>` {+mdb-server+} manual page.

The object on which you call the ``watch()`` method on determines the scope of
events that the change stream listens for:

- ``MongoCollection.watch()`` monitors a collection.
- ``MongoDatabase.watch()`` monitors all collections in a database.
- ``MongoClient.watch()`` monitors all changes in the connected MongoDB deployment.

Filter the Events
~~~~~~~~~~~~~~~~~

The ``watch()`` method optionally takes an **aggregation pipeline**  which
consists of an array of **stages** as the first parameter to filter and
transform the change event output as follows:

.. code-block:: java

   List<Bson> pipeline = Arrays.asList(
                           Aggregates.match(
                              Filters.in("operationType",

>                                     Arrays.asList("insert", "update"))),

Author: Replaced all instances of `Array.asList()`

                           Aggregates.match(
                              Filters.lt("fullDocument.runtime", 15)));
   ChangeStreamIterable<Document> changeStream = database.watch(pipeline);

.. note::

   For update operation change events, change streams only return the modified
   fields by default rather than the entire updated document. You can configure
   your change stream to also return the most current version of the document
   by calling the ``fullDocument()`` member method of the ``ChangeStreamIterable``
   object with the value ``FullDocument.UPDATE_LOOKUP`` as follows:

Manage the Output
~~~~~~~~~~~~~~~~~

> The ``watch()`` method returns an instance of ``ChangeStreamIterable``, a class

Reviewer: ```suggestion
The ``watch()`` method returns an instance of ``ChangeStreamIterable``, an interface
```

that offers several methods to access, organize, and traverse the results.

> ``ChangeStreamIterable`` also inherits methods from its parent class,

Reviewer: ```suggestion
``ChangeStreamIterable`` also inherits methods from its parent interface,
```

``MongoIterable`` which implements the core Java interface ``Iterable``.

You can call ``forEach()`` on the ``ChangeStreamIterable`` to handle
events as they occur, or you can use the ``iterator()`` method which

> returns a ``MongoCursor`` instance that you can use to traverse the results.

Reviewer: ```suggestion
returns a ``MongoChangeStreamCursor`` instance that you can use to traverse the results.
```

.. important:: forEach() blocks the current thread

>    Calls to ``forEach()`` block the current thread while the

Reviewer: This advice applies equally to iterating the change stream via a cursor.  Not sure why you are calling it out here.  You might say that you can interleave other work between calls to `next` on the cursor, but the same is true in the body of the lambda that the application passes to `forEach`

   corresponding change stream listens for events. If your program
   needs to continue executing other logic, such as processing requests or
   responding to user input, consider creating and listening to your
   change stream in a separate thread.

> You can call the following methods on the ``MongoCursor`:

Reviewer: ```suggestion
You can call the following methods on the ``MongoChangeStreamCursor`:
```

 - ``hasNext()``: checks if there are more results. 
 - ``next()`` returns the next document in the collection.
 - ``tryNext()`` immediately returns eitherthe next available element in the
   change stream or ``null``. 

> Unlike the ``MongoCursor`` returned by other queries, a ``MongoCursor`` associated

Reviewer: ```suggestion
Unlike the ``MongoCursor`` returned by other queries, a ``MongoChangeStreamCursor`` associated
```

   see the :ref:`<replica-set-oplog>` {+mdb-server+} manual page.

The object on which you call the ``watch()`` method on determines the scope of
events that the change stream listens for:

- ``MongoCollection.watch()`` monitors a collection.
- ``MongoDatabase.watch()`` monitors all collections in a database.
- ``MongoClient.watch()`` monitors all changes in the connected MongoDB deployment.

Filter the Events
~~~~~~~~~~~~~~~~~

> The ``watch()`` method optionally takes an **aggregation pipeline**  which

Reviewer: This still doesn't read well to my ears.  To many "which"s.  Can you try to re-word it?

      ...
   }

.. _java-usage-watch:

Watch Example: Full Files 
~~~~~~~~~~~~~~~~~~~~~~~~~

.. include:: /includes/crud/example-intro.rst

This example demonstrates how to open a change stream by using the watch method.
The ``Watch.java`` file calls the ``watch()`` method with a pipeline as an
argument to filter for only ``"insert"`` and ``"update"`` events. The
``WatchCompanion.java`` file inserts, updates and deletes a document.

To use the following examples, run the files in this order:

#. Run the ``Watch.java`` file.
#. Run the ``WatchCompanion.java`` file.

.. note::

   The ``Watch.java`` file will continue running until the
   ``WatchCompanion.java`` file is run.

``Watch.java``:

.. literalinclude:: /includes/crud/Watch.java
   :language: java

``WatchCompanion.java``:

.. literalinclude:: /includes/crud/WatchCompanion.java
   :language: java

Full File Example Output
````````````````````````

The preceding applications will generate the following output:

> ``Watch.java`` will capture on the  ``insert`` and ``update`` operations are

Reviewer: This sentence is not grammatically correct, I think.
